content/src/controller.js

```diff
@@ -786,6 +793,87 @@ function saveAction(forceOverwrite, loginPrompt, doneCallback) {
   });
 }
 
+function generateThumbnailDataUrl() {
+  var THUMBNAIL_SIZE = 128;
+  // Use the same background as the default icons.
+  var BACKGROUND_COLOR = '#eeeeee';
+
+  // Get the canvas inside the iframe.
+  var iframe = document.getElementsByTagName('iframe')[0];
+  var innerDoc = iframe.contentDocument || iframe.contentWindow.document;
+  var canvas = innerDoc.getElementsByTagName('canvas')[0];
+  var w = canvas.width;
+  var h = canvas.height;
+
+  if (canvas.id === 'turtle') {
+    return;
+  }
+
+  // Get the image data.
+  var ctx = canvas.getContext('2d');
+  var imageData = ctx.getImageData(0, 0, w, h);
+
+  // Initialize the coordinates for the image region,
+  // topLeft is initialized to bottom right,
+  // and bottomRight is initialized to top left.
+  var topLeft = { x: h, y: w };
+  var bottomRight = { x: 0, y: 0 };
+
+  // Iterate through all the points to find the "interesting" region.
+  var x, y, index;
+  for (y = 0; y < h; y++) {
+    for (x = 0; x < w; x++) {
+      // Every pixel takes up 4 slots in the array, contains R, G, B, A.
+      index = (y * w + x) * 4;
+      // Thus `index + 3` is the index of the Alpha value.
+      if (imageData.data[index + 3] > 0) {
+        if (x < topLeft.x) {
+          topLeft.x = x;
+        }
+        if (x > bottomRight.x) {
+          bottomRight.x = x;
+        }
+        if (y < topLeft.y) {
+          topLeft.y = y;
+        }
+        if (y > bottomRight.y) {
+          bottomRight.y = y;
+        }
+      }
+    }
+  }
+
+  // Calculate the actual image size.
+  var imageWidth = bottomRight.x - topLeft.x + 1;
+  var imageHeight = bottomRight.y - topLeft.y + 1;
+
+  // Find the longer edge and make it a square.
+  var longerEdge;
+  if (imageWidth > imageHeight) {
+    longerEdge = imageWidth;
+    topLeft.y -= (imageWidth - imageHeight) / 2;
+  } else {
+    longerEdge = imageHeight;
+    topLeft.x -= (imageHeight - imageWidth) / 2;
+  }
+
+  // Draw the cropped image in a temp canvas and scale it down.
+  var tempCanvas = document.createElement('canvas');
+  var tempCanvasCtx = tempCanvas.getContext('2d');
+  tempCanvas.width = THUMBNAIL_SIZE;
+  tempCanvas.height = THUMBNAIL_SIZE;
+  tempCanvasCtx.fillStyle = BACKGROUND_COLOR;
```

Comment:
Yes that sounds better. I wanted to make it the same background color as the default thumbnails. But now I think of a better way to do it: I shall rather do it with css.